refactor: make `frappe.db.bulk_insert` work as expected

pruthvi145 opened the pull request:
## Issue

`frappe.db.bulk_insert` is not working as expected:
 - It will not insert any row if there are less than 3 values
 - It will not add 1st row at all. e.g if I'm adding 5 values, it will only add 4.
 - It will add values one by one after 2 values, instead it should have inserted items (in db) in chunk of 10000 (as per the code written before).

## Changes Made
 - Solved above issues 
 - use better way to chunk list
 - Added Postgres support for bulk_insert API
 
And now `bulk_insert` will only do **1 db call for each 10000** values.

Note: For testing purpose I made `Test Bulk Insert` doctype and keep chunk size of 100.

## Before
![image](https://user-images.githubusercontent.com/43115036/161979365-c1100745-7602-47d2-a9b8-62d797d2039f.png)

## After
![image](https://user-images.githubusercontent.com/43115036/161978344-3c17d56b-2195-40f4-b00c-e9478d4083f1.png)

nodocs

pruthvi145 committed:
fix: use better way to chunk list

pruthvi145 committed:
Merge branch 'develop' into fix-bulk-insert

pruthvi145 committed:
refactor: use qb

pruthvi145 committed:
feat: chunk_size as keyword-only argument

pruthvi145 committed:
Merge branch 'develop' into fix-bulk-insert

gavindsouza commented on frappe/database/database.py:
```diff
@@ -1228,31 +1228,30 @@ def log_touched_tables(self, query, values=None):
+		values,
+		ignore_duplicates=False,
+		*,
+		chunk_size=10000
```
```suggestion
		chunk_size=10_000
```

Just a bit easier to read 😄

gavindsouza reviewed (changes requested):
Can you install `pre-commit` so that your code is styled before you push here? (And the Linters' build passes 😆)

Also, there aren't any tests for bulk_insert, can you add some to validate your fix? (One test for the cases you mentioned and another to test the chunking should do it)

Else all, LGTM. Thanks for the fix :)

pruthvi145 committed:
fix: formatting issues

pruthvi145 committed:
fix: table initialization outside loop

lapardnemihk1099 committed:
test: bulk_insert

pruthvi145 committed:
Merge branch 'develop' into fix-bulk-insert

lapardnemihk1099 commented:
# Test screenshot:

## Before
![Screenshot-2022-04-18-130130](https://user-images.githubusercontent.com/84378369/163777968-f34a470d-f5e1-4a9b-8fed-0811b9682ab2.png)

## After
![Screenshot-2022-04-18-133752](https://user-images.githubusercontent.com/84378369/163777950-53ae78fe-77fa-4ddf-9ef3-e9cebde418ce.png)

pruthvi145 committed:
fix: delete instead of rollback

pruthvi145 committed:
chore: add ignore_duplicates argument

pruthvi145 committed:
fix: ignore_duplicates for Postgres

pruthvi145 committed:
Merge branch 'develop' into fix-bulk-insert

gavindsouza committed:
test: Update delete refs for test_bulk_insert

Delete only files that were generated by this test

gavindsouza committed:
fix: Use string comparison instead of db_type_is Enum

Planning on making this check more succint in a separate PR. Using a
pattern that's common in the codebase for ease of mass replacement.
